src/nix/path-info.cc

```diff
@@ -86,14 +94,58 @@ struct CmdPathInfo : StorePathsCommand, MixJSON
         for (auto & storePath : storePaths)
             pathLen = std::max(pathLen, store->printStorePath(storePath).size());
 
+        StorePathSet substitutablePaths;
+        if (showSubStatus) {
+            StringSet pathsS;
+            // query each configured substituter
+            for (auto & sub : getDefaultSubstituters()) {
+              if (sub->storeDir != store->storeDir) continue;
+              if (!sub->wantMassQuery) continue;
+
+              auto paths = sub->queryValidPaths(StorePathSet(storePaths.begin(), storePaths.end()));
+              for (auto & path : paths) {
+                pathsS.emplace(store->printStorePath(path));
+              }
+            }
+
+            for (auto & str : pathsS) {
+              substitutablePaths.emplace(store->followLinksToStorePath(str));
+            }
+        }
+
         if (json) {
-            std::cout << store->pathInfoToJSON(
+            auto json = store->pathInfoToJSON(
                 // FIXME: preserve order?
                 StorePathSet(storePaths.begin(), storePaths.end()),
-                true, showClosureSize, SRI, AllowInvalid).dump();
+                true, showClosureSize, SRI, AllowInvalid);
+
+            if (showSubStatus) {
+                auto jsonNew = nlohmann::json::array();
+
+                for (auto v : json) {
+                    std::string const storePathS = v["path"];
+                    substitutablePaths.find(store->followLinksToStorePath(storePathS)) ==
+                            substitutablePaths.end()
+                        ? v["substitutable"] = false
+                        : v["substitutable"] = true;
```

Comment:
```suggestion
                    v["substitutable"] = (bool) substitutablePaths.count(store->followLinksToStorePath(storePathS));
```

Reply:
Maybe instead of a `substitutable` bool, it should return an array of substituters?

Reply:
Good idea

Reply:
So now that I've figured out why `Store::querySubstitutablePaths` wasn't working: https://github.com/NixOS/nix/pull/7587#discussion_r1068472573, I'm not sure how to easily do this. If I stick with defining the logic myself it should be easy though.

Attempts at using `querySubstitutablePathInfos` result in much slower queries which is why I moved away from that implementation. One possibility would be to add a flag to `queryMissing` to ignore the local store, but that would break the API.